packages/nimble/inst/CppCode/dists.cpp

```diff
@@ -271,6 +279,241 @@ SEXP C_rwish_chol(SEXP chol, SEXP df, SEXP scale_param)
+  char sideL('L');
+  char diag('N');
+  char transT('T');
+  char transN('N');
+  double alpha(1.0);
+  double beta(0.0);
+
+  // calculations here generate W ~ Wi(S) for scale_param = 1 and W ~ Wi(S^-1) for scale_param = 0,
+  // and then invert the result, giving InvWi(S^-1) for scale_param = 1 and InvWi(S) for scale_param = 0,
+  // because scale for Wishart is inverse of scale for inverse-Wishart
+  // therefore generate under opposite parameterization
+  scale_param = 1 - scale_param;  
+
+
+  double* cholCopy;
+  int i, j, uind, lind;
+  
+#ifdef IEEE_754
+  if (R_isnancpp(chol, p*p) || R_isnancpp(df) || R_isnancpp(scale_param)) {
+    for(j = 0; j < p*p; j++) 
+      Z[j] = R_NaN;
+    return;
+  }
+#endif
+    
+  // also covers df < 0
+  if(df < (double) p) {
+    for(j = 0; j < p*p; j++) 
+      Z[j] = R_NaN;
+    return;
+  }
+  
+  // fill diags with sqrts of chi-squares and upper triangle (for scale_param) with std normals - crossproduct of result is standardized Wishart; based on rWishart in stats package
+  for(j = 0; j < p; j++) {
+    Z[j*p + j] = sqrt(rchisq(df - (double) j)); 
+    for(i = 0; i < j; i++) {
+      uind = i + j * p, /* upper triangle index */
+      lind = j + i * p; /* lower triangle index */
+      Z[(scale_param ? uind : lind)] = norm_rand();
+      Z[(scale_param ? lind : uind)] = 0;
+    }
+  }
+ 
+  if(overwrite_inputs)
+    cholCopy = chol;
+  else {
+    cholCopy = new double[p*p];
+    for(i = 0; i < p*p; i++) 
+      cholCopy[i] = chol[i];
+  }
+
+  // with S parameterization: t(chol)%*%t(Z)%*%Z%*%chol is Wishart-distributed, so take inverse based on pieces
+  // with R parameterization: inverse(chol)%*%Z%*%t(Z)%*%inverse(t(chol)) is Wishart-distributed, again inverse of pieces
+  if(scale_param) {
+    // Z %*% chol
+    F77_CALL(dtrmm)(&sideL, &uploU, &transN, &diag, &p, &p, &alpha, Z, &p, cholCopy, &p);
+    double* iden = new double[p*p];
+    for(j = 0; j < p; j++)
+      for(i = 0; i < p; i++)
+        if(i == j) iden[j*p+i] = 1.0; else iden[j*p+i] = 0.0;
+    // inverse(Z %*% chol)
+    F77_CALL(dtrsm)(&sideL, &uploU, &transN, &diag, &p, &p, &alpha, cholCopy, &p, iden, &p);
+    // crossproduct of result
+    F77_CALL(dgemm)(&transN, &transT, &p, &p, &p, &alpha, iden, &p, iden, &p, &beta, Z, &p);    
+    delete [] iden;
+  } else {    
+    // solve(Z, chol)
+    F77_CALL(dtrsm)(&sideL, &uploL, &transN, &diag, &p, &p, &alpha, Z, &p, cholCopy, &p);
+    // crossproduct of result
+    F77_CALL(dgemm)(&transT, &transN, &p, &p, &p, &alpha, cholCopy, &p, cholCopy, &p, &beta, Z, &p);
+  }
+
+  if(!overwrite_inputs)
+    delete [] cholCopy;
+}
+
+SEXP C_rinvwish_chol(SEXP chol, SEXP df, SEXP scale_param) 
+// generates single Inverse Wishart draw given Cholesky of scale or rate matrix
+// Cholesky matrix should be given as a numeric vector in column-major order
+//   including all n x n elements; lower-triangular elements are ignored
+{
+  if(!isReal(chol) || !isReal(df) || !isReal(scale_param))
```

Comment:
I haven't followed the whole discussion in terms of revision of our error trapping. If NIM_ASSERT is what we've agreed on, then this is fine by me.

I would like to have a bit of discussion amongst all of us in-person as to the plan for error trapping going forward so we all know and are consistent (and ideally go back and revise current error trapping).

Fritz, perhaps you can remind us to discuss this next Monday?